### PR TITLE
Add Ik-cca encryption scheme

### PR DESCRIPTION
**Pull request by riemann89**
This PR provides an ik-cca encryption scheme for Zeth (see #2)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ To use the Zeth functionality, 3 components are required:
 
 We use 3 terminals, one for each of the above components.
 
+Note: Mac users should increase docker runtime memory from 2gb to 4gb to make Terminal 1 running.
```
Comment (dtebbs) [inline README.md:30]: `make Terminal 1 running` -> `allow Terminal 1 to complete successfully`

File: pyClient/commands/utils.py [hunk 2]
```diff
@@ -192,7 +192,7 @@ def _do_sync() -> int:
                 new_merkle_root = mix_result.new_merkle_root
 
                 for note_desc in wallet.receive_notes(
-                        mix_result.output_events, mix_result.sender_k_pk):
+                        mix_result.output_events):
```
Comment (dtebbs) [inline pyClient/commands/utils.py:195]: (fits on one line now)

File: pyClient/test/test_merkle_tree.py [hunk 3]
```diff
@@ -30,7 +30,7 @@ def setUpClass() -> None:
     @staticmethod
     def tearDownClass() -> None:
         if exists(MERKLE_TREE_TEST_DIR):
-            print(f"Removing test dir: {MERKLE_TREE_TEST_DIR}")
+            # Removing test dir: {MERKLE_TREE_TEST_DIR}"
```
Comment (dtebbs) [inline pyClient/test/test_merkle_tree.py:33]: I think you can just remove this line

File: pyClient/zeth/encryption.py [hunk 4]
```diff
@@ -4,21 +4,36 @@
 #
 # SPDX-License-Identifier: LGPL-3.0+
 
-from nacl.public import PrivateKey, PublicKey  # type: ignore
-import nacl.encoding  # type: ignore
 from typing import NewType
 
-# Thin wrapper around the nacl PublicKey and PrivateKey.  Note nacl doesn't
-# include much type information, so we try to enforce strict types.
+from cryptography.hazmat.primitives.asymmetric.x25519 \
```
Comment (dtebbs) [inline pyClient/zeth/encryption.py:9]: Just a small thing, but we try to use the convention of ordering imports: local (zeth), external libraries, standard lib (except __future__ which must be first). Would you mind reordering here to be consistent?

Comment (riemann89) [reply pyClient/zeth/encryption.py:9]: I have switched them. Sorry, what do you mean with "future"?

File: pyClient/test/test_merkle_tree.py (hunk 3 above)
Comment (riemann89) [reply pyClient/test/test_merkle_tree.py:33]: Done

File: pyClient/commands/utils.py (hunk 2 above)
Comment (riemann89) [reply pyClient/commands/utils.py]: Changed

File: README.md (hunk 1 above)
Comment (riemann89) [reply README.md]: Changed

**Comment by AntoineRondelet** [comment]
@riemann89 please make sure to rebase your branch on top of develop since the build has been repaired.

File: pyClient/test/test_zeth_constants.py [hunk 5]
```diff
@@ -4,39 +4,44 @@
...
+from zeth.constants import APK_LENGTH, HIDDEN_VALUE_LENGTH,\
+    RHO_LENGTH, TRAPR_LENGTH
 
 
 class TestZethConstants(TestCase):
 
     def test_encrypt_decrypt(self) -> None:
         """
         Tests the correct encrypt-decrypt flow: decrypt(encrypt(m)) == m
+        where m is 120 bytes long.
```
Comment (AntoineRondelet) [inline pyClient/test/test_zeth_constants.py:21]: ` 120 bytes long` ->  120-byte long

I'd rather write: `where m is encoded on (APK_LENGTH + HIDDEN_VALUE_LENGTH + RHO_LENGTH +TRAPR_LENGTH)/BYTE_LEN bytes`

File: pyClient/zeth/constants.py [hunk 6]
```diff
@@ -51,3 +51,25 @@
 
 # Seed for MIMC
 MIMC_MT_SEED: str = "clearmatics_mt_seed"
+
+# Key Derivation Tag "ZethEnc" utf-8 encoding
+KDF_TAG: bytes = b'ZethEnc'
+
+# Note constants
+APK_LENGTH: int = 256
+HIDDEN_VALUE_LENGTH: int = 64
```
Comment (AntoineRondelet) [inline pyClient/zeth/constants.py:60]: `HIDDEN_VALUE_LENGTH` -> `NOTE_VALUE_LENGTH`

File: pyClient/zeth/encryption.py (hunk 4 above)
Comment (AntoineRondelet) [reply pyClient/zeth/encryption.py:13]: @riemann89 https://github.com/clearmatics/zeth/blob/develop/pyClient/zeth/contracts.py#L7

File: pyClient/test/test_zeth_constants.py (hunk 5 above)
Comment (riemann89) [reply pyClient/test/test_zeth_constants.py:21]: Changed

File: pyClient/zeth/constants.py (hunk 6 above)
Comment (riemann89) [reply pyClient/zeth/constants.py:94]: Changed

**Comment by riemann89** [comment]
Just rebased on develop

File: pyClient/zeth/mixer_client.py [hunk 7]
```diff
@@ -657,9 +673,11 @@ def compute_h_sig(
 
 def trap_r_randomness() -> str:
     """
-    Compute randomness "r" as 48 random bytes
+    Compute randomness "r" as 32 random bytes
     """
-    return bytes(Random.get_random_bytes(32)).hex()
+    assert (constants.TRAPR_LENGTH & 7) == 0
+    trapr_len_bytes = constants.TRAPR_LENGTH >> 3
+    return bytes(Random.get_random_bytes(trapr_len_bytes)).hex()
```
Comment (dtebbs) [inline pyClient/zeth/mixer_client.py:680]: Here I used constants when fixing some issues during rebase.  However, there seems to be several other places where constants can be used.  #203

**Pull request merged**